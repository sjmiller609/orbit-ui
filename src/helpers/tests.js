import { BrowserRouter } from 'react-router-dom'
import { shape } from 'prop-types'
import { mount } from 'enzyme'

// Instantiate router context
const router = {
  history: new BrowserRouter().history,
  route: {
    location: {},
    match: {},
  },
}

const createContext = () => ({
  context: { router },
  childContextTypes: { router: shape({}) },
})

// wrap with router context - to mount RR links correctly
export function mountWrap(node) {
  return mount(node, createContext())
}
