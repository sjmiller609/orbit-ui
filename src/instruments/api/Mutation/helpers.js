const has = (str, find) => ~str.toLowerCase().indexOf(find.toLowerCase())

export const formErrors = ({ error, errors }) => {
  if (!error) return

  let err = JSON.stringify(error).toLowerCase()
  return errors.find(e => {
    if (e.key && has(err, e.key)) return e
    else if (e.allKeys && e.allKeys.every(k => has(err, k))) return e
    else if (e.someKeys && e.someKeys.some(k => has(err, k))) return e
  })
}
