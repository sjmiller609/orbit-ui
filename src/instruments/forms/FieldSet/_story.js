import React from 'react'
import { storiesOf } from '@storybook/react'
import { FieldSet } from 'instruments'

storiesOf('Instruments|Forms.FieldSet', module).add('Default', () => (
  <FieldSet name="mock" onChange={() => null} updateErrors={() => null} />
))
