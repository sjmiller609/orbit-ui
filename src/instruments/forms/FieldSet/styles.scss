@import 'src/styles/mixins.scss';

.remove {
  margin-bottom: 2.2rem;
  margin-left: auto;
  width: auto;
}
