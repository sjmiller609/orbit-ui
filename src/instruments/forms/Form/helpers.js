// faltten and stringify nested keys
export const unpack = (obj, ns, hash) => {
  if (!obj) return
  return Object.entries(obj).reduce((acc, [key, value]) => {
    const keypath = ns ? `${ns}.${key}` : key
    if (value && typeof value === 'object') {
      return unpack(value, keypath, acc)
    }
    acc[keypath] = value
    return acc
  }, hash || {})
}

const roll = (keys, obj, value) => {
  const key = keys[0]
  // case 1 - not obj
  if (keys.length === 1) {
    obj[key] = value
    return obj
  }
  // case 2 - doesn't exist
  if (!obj[key]) {
    obj[key] = isNaN(keys[1]) ? {} : []
  }
  // case 3 - exists
  obj[key] = roll(
    keys.slice(1),
    Array.isArray(obj[key]) ? [...obj[key]] : { ...obj[key] },
    value
  )

  return obj
}

// convert back into obj
export const pack = (obj, type = {}) => {
  return Object.entries(obj).reduce((acc, [key, value]) => {
    const keys = key.split('.')

    return roll(keys, acc, value)
  }, type)
}

export const packChild = ({ name, obj }) => {
  const obj2 = {}
  let isArray
  Object.entries(obj).forEach(k => {
    if (name.length < k[0].length && k[0].indexOf(name) === 0) {
      const n = k[0].slice(name.length + 1)
      // must be true for all matching
      isArray = !isNaN(n.slice(0, n.indexOf('.')))
      obj2[n] = k[1]
    }
  })
  if (Object.keys(obj2).length) return pack(obj2, isArray ? [] : {})
  return null
}

export const removeChild = ({ name, obj = {} }) => {
  const obj2 = {}
  Object.entries(obj).forEach(k => {
    if (name.length >= k[0].length || k[0].indexOf(name) !== 0) {
      obj2[k[0]] = k[1]
    }
  })
  return obj2
}
