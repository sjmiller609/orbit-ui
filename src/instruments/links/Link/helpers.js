// check if url is external
export const externalUrl = url => {
  if (typeof url === 'object') return false
  const path = /^https?:\/\/|^\/\//i
  if (path.test(url) && !~url.indexOf(window.location.origin)) return true
  return
}
