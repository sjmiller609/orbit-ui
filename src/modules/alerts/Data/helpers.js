import { formErrors } from 'instruments'

const errors = [
  {
    key: 'permissions',
    name: 'properties.alert_emails',
    error: 'Contact your workspace admin to upgrade your permissions.',
  },
]

export const handleError = error => formErrors({ error, errors })

export const trimError = error => error.split(':')[1].trim()
