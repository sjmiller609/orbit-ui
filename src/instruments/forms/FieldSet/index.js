import PropTypes from 'prop-types'
import React from 'react'
import s from './styles.scss'
import classnames from 'classnames'
import { Field, TextField, TextButton, Button, Box } from 'instruments'
import { jsonEqual } from 'helpers/compare'

class FieldSet extends React.Component {
  validate = this.validate.bind(this)
  remove = this.remove.bind(this)
  add = this.add.bind(this)
  renderField = this.renderField.bind(this)
  fieldProps = this.fieldProps.bind(this)
  removeEmpties = this.removeEmpties.bind(this)

  componentWillMount() {
    const { value } = this.props
    if (!value || !value.length) this.add()
  }

  componentDidMount() {
    const { value, registerOnSubmit, name } = this.props
    this.validate(value) // adds required fields to form
    if (registerOnSubmit)
      registerOnSubmit({ name, onSubmit: this.removeEmpties })
  }

  componentWillReceiveProps({ value }) {
    if (!jsonEqual(value, this.props.value)) {
      // run validation
      this.validate(value)
    }
  }

  removeEmpties(data) {
    if (!data) return
    return data.filter(d => !!d)
  }

  validate(value) {
    const { name, validate, updateErrors } = this.props
    let e
    if (validate) {
      // only validation is if passed in as prop
      e = validate(value)
      updateErrors(name, e)
    }
  }

  fieldProps(i) {
    const { className, ...props } = this.props.fieldProps
    return {
      ...props,
      ...this.props.formField(this.props.name + '.' + i),
      className: classnames(s.field, className),
      required: this.props.required,
      formField: this.props.formField,
    }
  }

  remove(i) {
    const { onChange, value } = this.props
    const v2 = Array.from(value)
    v2.splice(i, 1)
    onChange(null, v2)
  }

  add() {
    const { onChange, value } = this.props
    const v2 = Array.from(value)
    v2.push(null)
    onChange(null, v2)
  }

  // any empty fields must have the field name = next array item to prevent empties getting saved
  renderField(i) {
    const { FieldType } = this.props
    return (
      <Box align="flex-start" key={i}>
        <FieldType {...this.fieldProps(i)} />
        <TextButton
          className={s.remove}
          style="red"
          onClick={() => this.remove(i)}>
          Remove
        </TextButton>
      </Box>
    )
  }

  render() {
    const { value, id, className, title } = this.props
    const button = 'New' + (title ? ' ' + title : '')
    // const length = emptyKeys.length + value.length
    // let key = 0
    return (
      <div id={id} className={classnames(s.fields, className)}>
        {value.map((f, i) => this.renderField(i))}
        <Box>
          <Button className={s.add} style="outline" onClick={this.add}>
            {button}
          </Button>
        </Box>
      </div>
    )
  }
}

FieldSet.propTypes = {
  fieldProps: PropTypes.object,
  FieldType: PropTypes.func,
  formField: PropTypes.func,
  value: PropTypes.array,
  name: PropTypes.string.isRequired,
  id: PropTypes.string,
  validate: PropTypes.func,
  updateErrors: PropTypes.func,
  onChange: PropTypes.func,
  required: PropTypes.bool,
  className: PropTypes.string,
  title: PropTypes.string,
  registerOnSubmit: PropTypes.func,
}

FieldSet.defaultProps = {
  FieldType: TextField,
  fieldProps: {},
  value: [],
}

export default Field(FieldSet)
