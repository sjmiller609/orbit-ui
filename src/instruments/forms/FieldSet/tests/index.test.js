import React from 'react'
import { mountWrap } from 'helpers/tests'
import FieldSet from '../index'

const props = {
  fieldProps: {},
  FieldType: jest.fn(),
  formField: jest.fn(),
  value: [],
  name: 'test',
  id: '1234',
  validate: jest.fn(),
  updateErrors: jest.fn(),
  onChange: jest.fn(),
  required: true,
  className: 'testClassString',
  title: 'Test',
  registerOnSubmit: jest.fn(),
}

describe('FieldSet', () => {
  it('renders correctly', () => {
    const wrapper = mountWrap(<FieldSet {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})
