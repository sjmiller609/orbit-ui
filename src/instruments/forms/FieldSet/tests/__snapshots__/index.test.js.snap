// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldSet renders correctly 1`] = `
<Field
  FieldType={[Function]}
  className="testClassString"
  fieldProps={Object {}}
  formField={[Function]}
  id="1234"
  label=""
  name="test"
  onChange={[Function]}
  registerOnSubmit={[Function]}
  required={true}
  title="Test"
  updateErrors={[Function]}
  validate={[Function]}
  value={Array []}
>
  <FieldSet
    FieldType={[Function]}
    className="field required testClassString"
    error={null}
    fieldProps={Object {}}
    formField={[Function]}
    id="1234"
    label={null}
    name="test"
    onBlur={null}
    onChange={[Function]}
    registerOnSubmit={[Function]}
    required={true}
    setRef={[Function]}
    setShowError={[Function]}
    showError={false}
    title="Test"
    updateErrors={[Function]}
    validate={[Function]}
    value={Array []}
  >
    <div
      className="fields field required testClassString"
      id="1234"
    >
      <Box>
        <Flex
          flow="box"
        >
          <div
            className="flex box undefined"
          >
            <Button
              className="add"
              onClick={[Function]}
              style="outline"
            >
              <Link>
                <A
                  className="link"
                >
                  <a
                    className="link"
                  >
                    <button
                      className="button outline add"
                      onClick={[Function]}
                      type="button"
                    >
                      New Test
                    </button>
                  </a>
                </A>
              </Link>
            </Button>
          </div>
        </Flex>
      </Box>
    </div>
  </FieldSet>
</Field>
`;
